From: Cezary Wilmanski <cezary.wilmanski@adtran.com>
Date: Wed, 29 Oct 2025 11:32:12 +0100
Subject: [PATCH] net: dsa: mxl862xx: replace software reset by hardware reset via GPIO

It has been observed that when the MXL862xx switch firmware is upgraded, the
device can be left in a weird state where some ports don't function.

That's because the GPIO line is hooked up, but the driver does not support it.

To resolve that, the patch replaces a software reset by hardware reset via GPIO
and moves the procedure at the beginning of the MXL862xx switch setup.

Signed-off-by: Cezary Wilmanski <cezary.wilmanski@adtran.com>
---
 drivers/net/dsa/mxl862xx/mxl862xx.c |   37 ++++++++------------------
 drivers/net/dsa/mxl862xx/mxl862xx.h |    1 +
 2 files changed, 13 insertions(+), 25 deletions(-)

--- a/drivers/net/dsa/mxl862xx/mxl862xx.c
+++ b/drivers/net/dsa/mxl862xx/mxl862xx.c
@@ -10,6 +10,7 @@
 #include <linux/bitfield.h>
 #include <linux/delay.h>
 #include <linux/etherdevice.h>
+#include <linux/gpio/consumer.h>
 #include <linux/icmpv6.h>
 #include <linux/if_bridge.h>
 #include <linux/if_vlan.h>
@@ -4591,6 +4592,11 @@ static int mxl862xx_probe(struct mdio_de
 
 	dev_set_drvdata(dev, ds);
 
+	priv->reset = devm_gpiod_get_optional(dev, "reset", GPIOD_OUT_LOW);
+	if (IS_ERR(priv->reset))
+		return dev_err_probe(dev, PTR_ERR(priv->reset),
+				     "failed to request reset GPIO\n");
+
 	err = dsa_register_switch(ds);
 	if (err) {
 		set_bit(MXL862XX_FLAG_WORK_STOPPED, &priv->flags);
--- a/drivers/net/dsa/mxl862xx/mxl862xx.h
+++ b/drivers/net/dsa/mxl862xx/mxl862xx.h
@@ -5,6 +5,7 @@
 
 #include <asm/byteorder.h>
 #include <linux/bitops.h>
+#include <linux/gpio/consumer.h>
 #include <linux/mdio.h>
 #include <linux/mutex.h>
 #include <linux/workqueue.h>
@@ -415,6 +416,7 @@ struct mxl862xx_fw_version {
 struct mxl862xx_priv {
 	struct dsa_switch *ds;
 	struct mdio_device *mdiodev;
+	struct gpio_desc *reset;
 	struct work_struct crc_err_work;
 	struct work_struct rescue_heal_work;
 	unsigned long flags;
--- a/drivers/net/dsa/mxl862xx/mxl862xx-host.c
+++ b/drivers/net/dsa/mxl862xx/mxl862xx-host.c
@@ -10,6 +10,8 @@
 #include <linux/bitops.h>
 #include <linux/bits.h>
 #include <linux/crc16.h>
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
 #include <linux/iopoll.h>
 #include <linux/limits.h>
 #include <linux/unaligned.h>
@@ -498,9 +500,15 @@ int mxl862xx_reset(struct mxl862xx_priv
 {
 	int ret;
 
+	if (priv->reset) {
+		gpiod_set_value_cansleep(priv->reset, 1);
+		usleep_range(5000, 5100);
+		gpiod_set_value_cansleep(priv->reset, 0);
+		return 0;
+	}
+
 	mutex_lock_nested(&priv->mdiodev->bus->mdio_lock, MDIO_MUTEX_NESTED);
 
-	/* Software reset */
 	ret = mxl862xx_reg_write(priv, MXL862XX_MMD_REG_LEN_RET, 0);
 	if (ret)
 		goto out;
